### PR TITLE
crimson: recovery bug fix

### PR DESCRIPTION
**Pull request by xxhdx1985126**
Signed-off-by: Xuehan Xu <xxhdx1985126@gmail.com>
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by xxhdx1985126** [comment]
jenkins test make check

File: src/crimson/osd/pg_recovery.cc [hunk 1]
```diff
@@ -218,6 +218,9 @@ size_t PGRecovery::start_replica_recovery_ops(
 
       if (pg->get_recovery_backend()->is_recovering(soid)) {
 	logger().debug("{}: already recovering object {}", __func__, soid);
+	auto& recovery_waiter = pg->get_recovery_backend()->get_recovering(soid);
+	out->push_back(recovery_waiter.wait_for_recovered_blocking());
+	started++;
```
[the remark this answers is not in the document]
Comment (xxhdx1985126) [reply src/crimson/osd/pg_recovery.cc:223]: > i guess the reason why we need to take the already-started recovery into account is that `PGRecovery::start_recovery_ops()` should wait until **all** recoveries are completed? 

Yes, that's right.

> if so, could you amend the commit message to explain the rationale of this change?

Will do so:-)

File: src/crimson/osd/replicated_backend.cc [hunk 2]
```diff
@@ -90,10 +90,11 @@ ReplicatedBackend::_submit_transaction(std::set<pg_shard_t>&& pg_shards,
       }
       if (--peers->pending == 0) {
         peers->all_committed.set_value();
+	peers->all_committed = {};
```
[the remark this answers is not in the document]
Comment (xxhdx1985126) [reply src/crimson/osd/replicated_backend.cc:93]: ok:-)

File: src/crimson/osd/replicated_recovery_backend.cc [hunk 3]
```diff
@@ -401,8 +401,8 @@ seastar::future<ObjectRecoveryProgress> ReplicatedRecoveryBackend::build_push_op
     [this, &recovery_info, &progress, stat, pop]
     (auto& new_progress, auto& oi, auto& available, auto& v) {
     return [this, &recovery_info, &progress, &new_progress, &oi, pop, &v] {
+      v = recovery_info.version;
```
Comment (tchaikov) [inline src/crimson/osd/replicated_recovery_backend.cc:404]: could you explain the reason why we need this change in the commit message as well?

Comment (athanatos) [reply src/crimson/osd/replicated_recovery_backend.cc:404]: Yeah, I'm unclear on why you need to set it after the first one.

Comment (xxhdx1985126) [reply src/crimson/osd/replicated_recovery_backend.cc:404]: This is for scenarios where objects need more than one push to recover, in which later pushes will encapsulate everion 0'0 without this change. I'll add the explanation to the commit message:-)

Comment (athanatos) [reply src/crimson/osd/replicated_recovery_backend.cc:404]: Ah, recovery_info.version will be eversion_t() for the first pull, so you are using v here to represent oi.version or recovery_info.version depending on whether the pull recovery_info already has the version field filled in?

Comment (xxhdx1985126) [reply src/crimson/osd/replicated_recovery_backend.cc:404]: Yes, that's right

**Review by tchaikov — APPROVED** [review]

**Pull request merged**